Updated for SU2 v7.5.0 Blackbird

j-signorelli opened the pull request:
Adapter has been updated to work with SU2 Blackbird, and has been tested with Calculix using perpendicular-flap test case found in "tests" folder. Image below shows results, which visually matches results found [here](https://precice.org/tutorials-perpendicular-flap.html).

More verification tests likely to follow.

![PerpendicularFlap](https://user-images.githubusercontent.com/116454159/203864387-53ff4768-2854-4d5c-ab80-79078410b7dd.png)

MakisH commented on su2AdapterInstall:
```diff
@@ -51,38 +41,15 @@ cp ./adapter_files/precice.cpp $SU2_HOME/SU2_CFD/src || { printf >&2 "Cannot cop
 printf "\nPlease navigate to the SU2 home directory $SU2_HOME to configure and build SU2 as described in its installation description.\n"
 printf "Please enable MPI when configuring SU2.\n\n"
 
-# If user already configured SU2 before, print the command as reminder
-if [ -f $SU2_HOME/config.log ]
-then
-    printf "Last time you configured SU2 by invoking:\n"
-    sed '1,10!d' $SU2_HOME/config.log | grep -F "$ ./configure"
-    printf "You may want to reuse this command or adapt it to the following sample:\n"
-fi
-
-# Sample commands for configuring and making SU2
-printf "A sample SU2 configure procedure might look like: ./configure --prefix=\$SU2_HOME --enable-mpi\n"
-printf "A sample SU2 make procedure might look like: make -j 4\n\n"
-
-printf "Adapter installation completed. Make sure to follow the instructions above to build the adapted SU2.\n\n"
```
Since we are anyway using a specific version, let's document the instructions. I understand that they now are:

```bash
./meson.py build
./ninja -C build install
```

which automatically uses all cores.

j-signorelli replied:
No particular reason for removing it except now that SU2 uses meson and ninja, and I just haven't bothered taking the time yet to figure out where the previously run meson settings had been saved. It may be super straightforward.

Yes usually it seems most of the time the above lines correctly set up SU2 (and w/ MPI if it's available).

MakisH commented on tests/perpendicular-flap/README:
```diff
@@ -0,0 +1,5 @@
+ccx_preCICE -i flap -precice-participant Solid <-- NOT Calculix, must match yaml.cpp file format
+
+and
+
+SU2_CFD euler...
```
A test case! Nice! :heart_eyes: 

Maybe rename this to `README.md`, add a couple of sentences, wrap the code in code blocks, and refer to it from the main `README.md`.

Is this different than the tutorial case? It would make more sense to split this into two PRs: one with the code updates here, and one with the case file updates in the tutorials repository. The reference results can stay here for now, pointing to the tutorial.

One day we will replace this with out system tests, but this will still take a while :see_no_evil: As soon as we define the system for reference results and comparisons, I will move the results there.

j-signorelli replied:
Missed this, will do! This is same as the tutorial case but noted, let me know if/when would be good to submit a pull-request to the tutorials repo

MakisH replied:
You can already move the test case to a PR updating the tutorials repository, updating the current tutorial.

MakisH commented on su2AdapterInstall:
```diff
@@ -51,38 +41,15 @@ cp ./adapter_files/precice.cpp $SU2_HOME/SU2_CFD/src || { printf >&2 "Cannot cop
-autoreconf
+# Replace SU2_CFD/src/meson.build
+printf "Replacing original build related file..."
+#cp ./replacement_files/aclocal.m4 $SU2_HOME || { printf >&2 "Cannot copy replacement_files/aclocal.m4. Is variable SU2_HOME set correctly? Are you running the script from the correct directory?\nAborting.\n"; exit 1; }
+#cp ./replacement_files/configure.ac $SU2_HOME || { printf >&2 "Cannot copy replacement_files/configure.ac. Is variable SU2_HOME set correctly? Are you running the script from the correct directory?\nAborting.\n"; exit 1; }
```
These are not needed since v7, right? Maybe remove the comments, then.

j-signorelli replied:
Done 👍

MakisH commented on README.md:
```diff
@@ -68,7 +64,7 @@ The adapter is turned on or off via the native SU2 configuration file. If it is
-Moreover, in the SU2 configuration file grid movement must be allowed: `GRID_MOVEMENT= YES` and the type of grid movement must be set correctly for preCICE-coupled simulations: `GRID_MOVEMENT_KIND= PRECICE_MOVEMENT`. Also, the boundary, which is allowed to move needs to be specified. Here the name of the FSI-interface marker including its appending identifying number as stated above needs to be u ...
+Moreover, in the SU2 configuration file grid movement the type of grid movement must be set correctly for preCICE-coupled simulations: `SURFACE_MOVEMENT= PRECICE_MOVEMENT`. Also, the boundary, which is allowed to move needs to be specified. Here the name of the FSI-interface marker including its appending identifying number as stated above needs to be used, e.g., `MARKER_MOVING= ( wetSurface0 )`. ...
 
 ### Running the Adapted SU2 Executable
```
I cannot run `SU2_CFD` on my Ubuntu 22.04 (I can run the previous version, though):

```bash
~/repos/precice/tutorials/perpendicular-flap/fluid-su2 [develop]$ ./run.sh 
[tickly:159675] *** An error occurred in MPI_Win_create
[tickly:159675] *** reported by process [2347040769,0]
[tickly:159675] *** on communicator MPI_COMM_WORLD
[tickly:159675] *** MPI_ERR_WIN: invalid window
[tickly:159675] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[tickly:159675] ***    and potentially your MPI job)
~/repos/precice/tutorials/perpendicular-flap/fluid-su2 [develop]$ which SU2_CFD
/usr/local/bin/SU2_CFD
~/repos/precice/tutorials/perpendicular-flap/fluid-su2 [develop]$ ldd /usr/local/bin/SU2_CFD
	linux-vdso.so.1 (0x00007fff73fa1000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f37be7b8000)
	libmpi.so.40 => /lib/x86_64-linux-gnu/libmpi.so.40 (0x00007f37be681000)
	libmpi_cxx.so.40 => /lib/x86_64-linux-gnu/libmpi_cxx.so.40 (0x00007f37be667000)
	libprecice.so.2 => /home/gc/repos/precice/precice/build/libprecice.so.2 (0x00007f37be2aa000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f37be07e000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f37bdf97000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f37bdf77000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f37bdd4f000)
	libopen-rte.so.40 => /lib/x86_64-linux-gnu/libopen-rte.so.40 (0x00007f37bdc92000)
	libopen-pal.so.40 => /lib/x86_64-linux-gnu/libopen-pal.so.40 (0x00007f37bdbdf000)
	libhwloc.so.15 => /lib/x86_64-linux-gnu/libhwloc.so.15 (0x00007f37bdb81000)
	libboost_log_setup.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_log_setup.so.1.74.0 (0x00007f37bdae3000)
	libboost_program_options.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_program_options.so.1.74.0 (0x00007f37bda9e000)
	libxml2.so.2 => /lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f37bd8bc000)
	libpython3.10.so.1.0 => /lib/x86_64-linux-gnu/libpython3.10.so.1.0 (0x00007f37bd2e2000)
	libboost_log.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_log.so.1.74.0 (0x00007f37bd234000)
	libboost_filesystem.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_filesystem.so.1.74.0 (0x00007f37bd214000)
	libboost_thread.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_thread.so.1.74.0 (0x00007f37bd1f2000)
	libpetsc_real.so.3.15 => /usr/lib/petscdir/petsc3.15/x86_64-linux-gnu-real/lib/libpetsc_real.so.3.15 (0x00007f37bbe74000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f37bf998000)
	libevent_core-2.1.so.7 => /lib/x86_64-linux-gnu/libevent_core-2.1.so.7 (0x00007f37bbe3f000)
	libevent_pthreads-2.1.so.7 => /lib/x86_64-linux-gnu/libevent_pthreads-2.1.so.7 (0x00007f37bbe38000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f37bbe0e000)
	libboost_regex.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_regex.so.1.74.0 (0x00007f37bbd1b000)
	libicuuc.so.70 => /lib/x86_64-linux-gnu/libicuuc.so.70 (0x00007f37bbb20000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f37bbaf5000)
	libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007f37bbac2000)
	libHYPRE-2.22.1.so => /lib/x86_64-linux-gnu/libHYPRE-2.22.1.so (0x00007f37bb6aa000)
	libdmumps-5.4.so => /lib/x86_64-linux-gnu/libdmumps-5.4.so (0x00007f37bb496000)
	libscalapack-openmpi.so.2.1 => /lib/x86_64-linux-gnu/libscalapack-openmpi.so.2.1 (0x00007f37baeee000)
	libumfpack.so.5 => /lib/x86_64-linux-gnu/libumfpack.so.5 (0x00007f37bae42000)
	libamd.so.2 => /lib/x86_64-linux-gnu/libamd.so.2 (0x00007f37bae35000)
	libcholmod.so.3 => /lib/x86_64-linux-gnu/libcholmod.so.3 (0x00007f37bad54000)
	libklu.so.1 => /lib/x86_64-linux-gnu/libklu.so.1 (0x00007f37bad25000)
	libsuperlu.so.5 => /lib/x86_64-linux-gnu/libsuperlu.so.5 (0x00007f37bacb3000)
	libsuperlu_dist.so.7 => /lib/x86_64-linux-gnu/libsuperlu_dist.so.7 (0x00007f37bab4f000)
	libtrilinos_ml.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_ml.so.13.2 (0x00007f37ba896000)
	libfftw3.so.3 => /lib/x86_64-linux-gnu/libfftw3.so.3 (0x00007f37ba679000)
	libfftw3_mpi.so.3 => /lib/x86_64-linux-gnu/libfftw3_mpi.so.3 (0x00007f37ba65f000)
	liblapack.so.3 => /lib/x86_64-linux-gnu/liblapack.so.3 (0x00007f37b9f26000)
	libblas.so.3 => /lib/x86_64-linux-gnu/libblas.so.3 (0x00007f37b9e80000)
	libptscotch-6.1.so => /lib/x86_64-linux-gnu/libptscotch-6.1.so (0x00007f37b9dad000)
	libhdf5_openmpi.so.103 => /lib/x86_64-linux-gnu/libhdf5_openmpi.so.103 (0x00007f37b99f8000)
	libOpenCL.so.1 => /lib/x86_64-linux-gnu/libOpenCL.so.1 (0x00007f37b99e4000)
	libyaml-0.so.2 => /lib/x86_64-linux-gnu/libyaml-0.so.2 (0x00007f37b99c3000)
	libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x00007f37b9883000)
	libmpi_mpifh.so.40 => /lib/x86_64-linux-gnu/libmpi_mpifh.so.40 (0x00007f37b981d000)
	libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f37b9542000)
	libicui18n.so.70 => /lib/x86_64-linux-gnu/libicui18n.so.70 (0x00007f37b9211000)
	libicudata.so.70 => /lib/x86_64-linux-gnu/libicudata.so.70 (0x00007f37b75f3000)
	libmumps_common-5.4.so => /lib/x86_64-linux-gnu/libmumps_common-5.4.so (0x00007f37b7597000)
	libsuitesparseconfig.so.5 => /lib/x86_64-linux-gnu/libsuitesparseconfig.so.5 (0x00007f37b7592000)
	libcolamd.so.2 => /lib/x86_64-linux-gnu/libcolamd.so.2 (0x00007f37b7589000)
	libccolamd.so.2 => /lib/x86_64-linux-gnu/libccolamd.so.2 (0x00007f37b757a000)
	libcamd.so.2 => /lib/x86_64-linux-gnu/libcamd.so.2 (0x00007f37b756e000)
	libmetis.so.5 => /lib/x86_64-linux-gnu/libmetis.so.5 (0x00007f37b7504000)
	libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f37b74ba000)
	libbtf.so.1 => /lib/x86_64-linux-gnu/libbtf.so.1 (0x00007f37b74b4000)
	libptscotchparmetis-6.1.so => /lib/x86_64-linux-gnu/libptscotchparmetis-6.1.so (0x00007f37b74ab000)
	libCombBLAS.so.1.16.0 => /lib/x86_64-linux-gnu/libCombBLAS.so.1.16.0 (0x00007f37b7492000)
	libtrilinos_ifpack.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_ifpack.so.13.2 (0x00007f37b7300000)
	libtrilinos_amesos.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_amesos.so.13.2 (0x00007f37b72ab000)
	libtrilinos_galeri-epetra.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_galeri-epetra.so.13.2 (0x00007f37b7265000)
	libtrilinos_aztecoo.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_aztecoo.so.13.2 (0x00007f37b71df000)
	libtrilinos_zoltan.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_zoltan.so.13.2 (0x00007f37b70fd000)
	libtrilinos_epetraext.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_epetraext.so.13.2 (0x00007f37b6f95000)
	libscotch-6.1.so => /lib/x86_64-linux-gnu/libscotch-6.1.so (0x00007f37b6f06000)
	libtrilinos_epetra.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_epetra.so.13.2 (0x00007f37b6da3000)
	libtrilinos_teuchosparameterlist.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_teuchosparameterlist.so.13.2 (0x00007f37b6a82000)
	libtrilinos_teuchoscore.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_teuchoscore.so.13.2 (0x00007f37b6a25000)
	libptscotcherr-6.1.so => /lib/x86_64-linux-gnu/libptscotcherr-6.1.so (0x00007f37b6a20000)
	libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f37b6a0d000)
	libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3 (0x00007f37b65cb000)
	libcurl.so.4 => /lib/x86_64-linux-gnu/libcurl.so.4 (0x00007f37b6525000)
	libsz.so.2 => /lib/x86_64-linux-gnu/libsz.so.2 (0x00007f37b651e000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f37b64f4000)
	libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f37b64ac000)
	libpord-5.4.so => /lib/x86_64-linux-gnu/libpord-5.4.so (0x00007f37b6493000)
	libtrilinos_trilinosss.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_trilinosss.so.13.2 (0x00007f37b6463000)
	libtrilinos_teuchosremainder.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_teuchosremainder.so.13.2 (0x00007f37b6459000)
	libtrilinos_teuchosnumerics.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_teuchosnumerics.so.13.2 (0x00007f37b643c000)
	libtrilinos_teuchoscomm.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_teuchoscomm.so.13.2 (0x00007f37b6365000)
	libtrilinos_triutils.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_triutils.so.13.2 (0x00007f37b6307000)
	libscotcherr-6.1.so => /lib/x86_64-linux-gnu/libscotcherr-6.1.so (0x00007f37b6302000)
	libtrilinos_teuchosparser.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_teuchosparser.so.13.2 (0x00007f37b62a0000)
	libtrilinos_kokkoscore.so.13.2 => /lib/x86_64-linux-gnu/libtrilinos_kokkoscore.so.13.2 (0x00007f37b6216000)
	libnghttp2.so.14 => /lib/x86_64-linux-gnu/libnghttp2.so.14 (0x00007f37b61ec000)
	libidn2.so.0 => /lib/x86_64-linux-gnu/libidn2.so.0 (0x00007f37b61cb000)
	librtmp.so.1 => /lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f37b61ac000)
	libssh.so.4 => /lib/x86_64-linux-gnu/libssh.so.4 (0x00007f37b613d000)
	libpsl.so.5 => /lib/x86_64-linux-gnu/libpsl.so.5 (0x00007f37b6129000)
	libssl.so.3 => /lib/x86_64-linux-gnu/libssl.so.3 (0x00007f37b6085000)
	libgssapi_krb5.so.2 => /lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f37b6031000)
	libldap-2.5.so.0 => /lib/x86_64-linux-gnu/libldap-2.5.so.0 (0x00007f37b5fd2000)
	liblber-2.5.so.0 => /lib/x86_64-linux-gnu/liblber-2.5.so.0 (0x00007f37b5fc1000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f37b5ef0000)
	libbrotlidec.so.1 => /lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007f37b5ee2000)
	libaec.so.0 => /lib/x86_64-linux-gnu/libaec.so.0 (0x00007f37b5ed9000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007f37b5ed3000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f37b5ecb000)
	libunistring.so.2 => /lib/x86_64-linux-gnu/libunistring.so.2 (0x00007f37b5d1f000)
	libgnutls.so.30 => /lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f37b5b34000)
	libhogweed.so.6 => /lib/x86_64-linux-gnu/libhogweed.so.6 (0x00007f37b5aec000)
	libnettle.so.8 => /lib/x86_64-linux-gnu/libnettle.so.8 (0x00007f37b5aa6000)
	libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f37b5a24000)
	libkrb5.so.3 => /lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f37b5957000)
	libk5crypto.so.3 => /lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f37b5928000)
	libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f37b5922000)
	libkrb5support.so.0 => /lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f37b5914000)
	libsasl2.so.2 => /lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f37b58f9000)
	libbrotlicommon.so.1 => /lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007f37b58d6000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f37b58bc000)
	libp11-kit.so.0 => /lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f37b5781000)
	libtasn1.so.6 => /lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f37b5769000)
	libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f37b5762000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f37b574e000)
	libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007f37b573f000)
	libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007f37b5732000)
```

Any clue what I could be missing? This is not specific to the tutorial, I cannot start it in general.

j-signorelli replied:
Two things:

1. One thing to watch out for that SU2 does not successfully implement at least Intel MPI (https://www.cfd-online.com/Forums/su2-installation/228428-su2-parallel-install-build-source-failed.html#post776838) with default meson settings as you show above. I have had to run with the additional settings below to get it to work:

```
export CC=$(which mpicc)
export CXX=$(which mpicxx)
export MPICC=$(which mpicc)
export MPICXX=$(which mpicxx)

./meson.py build -Dcustom-mpi=true --prefix=...
./ninja -C build install
```

I am not sure how it does with other MPIs, but something similar may be needed. I've tried numerous different things that didn't involve needing to set `-Dcustom-mpi` but have never been successful.

I would try this above and see if it helps?

2. This is exactly the perpendicular flap tutorial case but a lot of names of options in the config file have changed since last SU2. The `tests` folder in this branch includes an updated config file that works for new SU2.

MakisH commented on replacement_files/CSinglezoneDriver.hpp:
```diff
@@ -0,0 +1,137 @@
+#pragma once
+#include "CDriver.hpp"
+
+// preCICE
+// This forward declaration of the Precice class is necessary to avoid a cyclic inclusion issue
+class Precice;
```
Is it not enough/better to include `precice.hpp`, instead?

j-signorelli replied:
This was something that I initially asked myself too (this same thing exists in the SU2 v6.0.0 adapter) and tried that, but if I remember correctly I *think* that the inclusion of `precice.hpp` led to continual cycles of inclusion ultimately leading back to this file from include statements in `precice.hpp`. I don't remember specifically though, but I do remember it caused an error.

MakisH replied:
I will look into that.

MakisH commented on adapter_files/precice.cpp:
```diff
@@ -478,6 +483,8 @@ double Precice::advance(double computedTimestepLength) {
       // the number of elapsed time steps; Achtung: ExtIter beginnt bei 0 (ohne Restart) und bei einem Restart
       // (Startlösung) nicht bei 0, sondern bei der Startiterationsnummer
+	  
+	  /* Temporarily commenting out - TODO
```
How temporarily? :innocent: 
Why?

j-signorelli replied:
Probably not something I will work on in the near future, might be useful keeping the old code lines there, but they definitely don't work for the newest version of SU2. Would you suggest I make that more clear or just remove it altogether?

MakisH reviewed (commented):
Thank you **so** much for contributing this! :hugs: :hugs: :hugs: 

I am not familiar enough with the code to review it in detail, but I already tried to build and run it. I have a couple of comments, and I need some help with actually running it, as it looks like I am getting an MPI-related issue (which I don't get with v6).

As a summary of the code changes, I understand that:
- SU2 moved from Automake to Meson. Some related files have been replaced, and the building instructions have changed.
- In `precice.cpp`, it looks like many objects are now wrapped in an additional layer:
```diff
- nDim(geometry_container[ZONE_0][MESH_0]->GetnDim()),
+ nDim(geometry_container[ZONE_0][INST_0][MESH_0]->GetnDim()),
```
which also leads to some quadruple (:fearful:) pointers.
- Regimes seem to be namespaced:
```diff
- bool incompressible = (config_container[ZONE_0]->GetKind_Regime() == INCOMPRESSIBLE);
+ bool incompressible = (config_container[ZONE_0]->GetKind_Regime() == ENUM_REGIME::INCOMPRESSIBLE);
```
- A type has changed from `double** Grad_PrimVar` to `CMatrixView<double> Grad_PrimVar`.

In `replacement_files`, there are a few new files. I have not looked in details regarding any preCICE-related changes, at least some seem to be marked by comments. I understand that SU2 renamed them. Maybe a `git mv` would help us understand the differences. Or did you start completely from scratch there?

There also seems to be a test case, which I guess is very similar (is it even identical?) to the tutorial. See the specific comment regarding this.

j-signorelli commented:
Hi,

Glad to help! Going to be honest (as you can probably tell): this is my first time submitting a pull-request / contributing to open source, so apologies if I'm not following the best "pull-request" etiquette. Here's a more detailed and precise (pun) list of all the code changes (also including the ones you mentioned):

* Most `replacement_files` have all been converted to their new file names + contents with preCICE code re-implemented:
    * `config_structure` -> `CConfig`
    * `driver_structure` -> `CSinglezoneDriver` (see note on this below)
    * `iteration_structure` -> `CIteration`

* Variables of interest below (in `precice.cpp`) are yes wrapped in an additional layer. This added layer is a time instance *only* used for Harmonic Balance (when `CHBDriver` being used). Otherwise just uses `INST_0`:

```c++
// Old:
- geometry_container[ZONE_0][MESH_0] = ...
// New:
+ geometry_container[ZONE_0][INST_0][MESH_0] = ...

// Old:
- solver_container[ZONE_0][MESH_0][FLOW_SOL] = ...
// New:
+ solver_container[ZONE_0][INST_0][MESH_0][FLOW_SOL] = ...

// Old:
- grid_movement[ZONE_0] = ...
// New:
+ grid_movement[ZONE_0][INST_0] = ...
```

* `CPoint` and `CVariable` previously represented data/geometrical information at a single node. Now, they contain data for the entire mesh. Numerous method calls in `CGeometry` and `CSolver` classes needed to be updated as a result.

* Newest version of SU2 now delineates unsteady mesh motion between `GRID_MOVEMENT` and `SURFACE_MOVEMENT`. `option_structure.hpp` has been updated so that `PRECICE_MOVEMENT` now lies under `SURFACE_MOVEMENT` (see below).  Updates were also made accordingly in what now is called `CIteration.cpp`, no longer `iteration_structure.cpp`

https://github.com/precice/su2-adapter/blob/f7a342ff09f74119cd242d09ed48850b5cb925c8/replacement_files/option_structure.hpp#L721-L758

* The `CDriver` class is now primarily virtual, where implementations of most of its methods are done in a variety of different child classes. I implemented preCICE capability specifically to the `CSinglezoneDriver` child class. Important to note that any discrete adjoint or multi-zone problems would not be compatible currently. Created `CSinglezoneDriver::Postprocessing` child function to de-allocate the additional preCICE vars before calling its super function `CDriver::Postprocessing`.

* New SU2 now uses meson and ninja, so instead of a Makefile update, a meson.build file was updated to include call to compile `precice.cpp` and include appropriate precice flags (cflags and lib flags)

* Previous adapter always saved and restored `GridVel_Grad`. New SU2 no longer even instantiates `GridVel_Grad` unless continuous adjoint problem being solved, which leads to problems when trying to access it. Adapter has been updated to check `CConfig::GetContinuous_Adjoint` before attempting to access or save it (otherwise segmentation fault).

* `CConfig::GetVolumetric_Movement` was updated to return `true` if `SURFACE_MOVEMENT= PRECICE_MOVEMENT`, ensuring that `CVolumetricMovement.StiffMatrix` is instantiated (needed for mesh deformation)

* In `Precice::advance`: `double** Grad_PrimVar` type was updated to match new returned SU2 class `CMatrixView<double>` that is returned from function call that sets it

* Config file for perpendicular flap test case was pulled directly from preCICE tutorials and updated to work for newest SU2 version; files stored in `tests/perpendicular-flap`

MakisH commented:
You are doing excellent so far! :smile: I will try running again and figuring out the cycling dependencies issue and come back to you something next week.

MakisH commented:
@j-signorelli I have not at all forgotten about this, I am just struggling a bit to find the time to look at it properly. Keep pushing any updates and I will look at it asap.

j-signorelli commented:
Hi @MakisH , just wanted to give you a heads up: this was actually my first time working with precice, and in hindsight I think that there are some improper implementations in the adapter here that I haven't updated (my guess is from an older version of precice): there is no initializeData option and writeBlockVectorData is called BEFORE readBlockVectorData, instead of being called prior to solver iteration.

It still works as I showed, but I have been working on what I believe is a better route: SU2 has a python wrapper that has a good amount of capabilities already implemented but not all required for coupling with precice. I've managed to add/edit JUST the python wrapper functions instead of any other aspect of the code, which is much less invasive and much faster to recompile. I've successfully gotten a CHT example (flow-over-heated-plate) working with that and the precice python bindings, and I'm pretty confident implementing FSI would be pretty simple.

j-signorelli commented:
Quick update on this: things are looking good! I am just working through a final bug (implicit coupling on >1 core) and then it should be good to go. I have attached images of 2-3 tutorial test cases (flow-over-heated-plate + flow-over-heated-plate-two-meshes, and perpendicular flap).

![image](https://user-images.githubusercontent.com/116454159/216749186-55b455c8-5127-46b4-8089-5e997c882398.png)
![image](https://user-images.githubusercontent.com/116454159/216749201-17dbfef3-1bb9-4907-9ab9-b89f5b9cc50e.png)

I am confident the discrepancy between the old and the new FSI case is because the new adapter I have been working on uses a new mesh deformation that SU2 is transitioning to. The image above I obtained sometime ago is now legacy. However further FSI and CHT verification cases should and will be completed.

When this is wrapped up, should I update the code in this pull request? Or just submit a new one altogether? I currently have been working on this on another repo.

kursatyurt commented:
> It still works as I showed, but I have been working on what I believe is a better route: SU2 has a python wrapper that has a good amount of capabilities already implemented but not all required for coupling with precice. I've managed to add/edit JUST the python wrapper functions instead of any other aspect of the code, which is much less invasive and much faster to recompile. I've successfully gotten a CHT example (flow-over-heated-plate) working with that and the precice python bindings, and I'm pretty confident implementing FSI would be pretty simple.

Hi, I had a chance to look at the python wrapper of SU2 and the CHT adapter you developed. It seems using the python API it is not possible to access, save and modify the old timesteps, hence it makes implicit coupling impossible right? Expect this the python adapters sound/seem really useful and easy to adapt.

j-signorelli commented:
Hi @kursatyurt , without additional edits that is correct. However: I have successfully added the necessary functions to the Python wrapper and variables to CDriver that make implicit coupling super easy in a re-formulated adapter. The repo I have been working on all this in is here: https://github.com/j-signorelli/su2-precice-pywrapper


I am actually planning on finalizing it today and updating the pull requests. The screenshots above do use implicit coupling with the adapted Python wrapper.

kursatyurt commented:
@j-signorelli  now it is clear, I just saw the initial work/repo. Your work looks awesome. For the wrapper mod and the python adapter(s), in my opinion, open another pull request as it can be confusing if you just add them to this PR. Small PRs are always easy to read for reviewers. 

Thank you very much for your contribution.

j-signorelli commented:
Hi @kursatyurt , haven't forgotten about this! I have been super busy. There appears to be some kind of bug with running perpendicular-flap in parallel - I took quite the deep dive in trying to figure it out last week but with no luck yet still. Running on 1 processes verses 3 processes gives quite different results. All data appears to be communicating correctly between ranks, and the script already has ensured no duplication of nodes are sent to preCICE either (halo nodes are handled properly).

MakisH commented:
Continuing in https://github.com/precice/su2-adapter/pull/32
